pkg/storegateway/indexheader/encoding/factory.go

```diff
@@ -154,3 +160,58 @@ func (df *DecbufFactory) CloseWithErrCapture(err *error, d Decbuf, format string
 
 	*err = merr.Err()
 }
+
+// filePool maintains a pool of file handles up to a maximum number, creating
+// new handles and closing them when required. get and put operations on this
+// pool never block. If there are no available file handles, one will be created
+// on get. If the pool is full, the file handle is closed on put.
+type filePool struct {
+	path    string
+	handles chan *os.File
+}
+
+// newFilePool creates a new file pool for path with cap capacity. If cap is 0,
+// get always opens new file handles and put always closes them immediately.
+func newFilePool(path string, cap uint) *filePool {
+	return &filePool{
+		path: path,
+		// We don't care if cap is 0 which means the channel will be unbuffered. Because
+		// we have default cases for reads and writes to the channel, we will always open
+		// new files and close file handles immediately if the channel is unbuffered.
+		handles: make(chan *os.File, cap),
+	}
+}
+
+// get returns a pooled file handle if available or opens a new one if there
+// are no pooled handles available.
+func (p *filePool) get() (*os.File, error) {
+	select {
+	case f := <-p.handles:
+		return f, nil
+	default:
+		return os.Open(p.path)
+	}
+}
+
+// put returns a file handle to the pool if there is space available or closes
+// the file handle if there is not.
+func (p *filePool) put(f *os.File) error {
+	select {
+	case p.handles <- f:
+		return nil
+	default:
+		return f.Close()
+	}
+}
```

Comment:
It would be nice to have a metric of how many new handles are opened and how many are being closed, as that would be useful to tune the pool size.

Reply:
Done in b17547b0d9bce05d438343cc627e3402d1d8ed1c